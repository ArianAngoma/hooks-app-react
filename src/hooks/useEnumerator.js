import {useState} from "react";

const useEnumerator = (initialState = 10) => {
    // Uso del hook useState para inicializar el state
    const [counter, setCounter] = useState(initialState);

    // Funcion para sumar al state
    const increment = () => {
        setCounter(counter + 1);
    }

    // Funcion poner el valor por defecto al state
    const reset = () => {
        setCounter(initialState);
    }

    // Funcion para restar al state
    const decrement = () => {
        setCounter(counter - 1);
    }

    return {
        counter,
        increment,
        reset,
        decrement
    }
}

export default useEnumerator;