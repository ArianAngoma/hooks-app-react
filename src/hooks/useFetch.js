import {useEffect, useRef, useState} from "react";

const useFetch = (url) => {
    const isMounted = useRef(true);

    const [state, setState] = useState({
        loading: true,
        error: null,
        data: null
    });

    // Solo lo harÃ¡ cuando el componente se carga por primera vez
    useEffect(() => {
        return () => {
            isMounted.current = false;
            // console.log(isMounted.current);
        }
    }, []);

    useEffect(() => {
        setState({data: null, loading: true, error: null});

        fetch(url).then(resp => resp.json()).then(data => {
            // console.log(isMounted.current);
            if (isMounted.current) {
                setState({
                    loading: false,
                    error: null,
                    data
                })
            }
        }).catch(() => {
            setState({
                data: null,
                loading: false,
                error: 'No se pudo cargar la info'
            })
        })
    }, [url]);

    return state;
}

export default useFetch;