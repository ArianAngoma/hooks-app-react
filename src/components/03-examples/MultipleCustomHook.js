import useFetch from "../../hooks/useFetch";
import useEnumerator from "../../hooks/useEnumerator";
import '../02-useEffect/effects.css';

const MultipleCustomHook = () => {
    // Llamamos al custom hook useEnumerator para hacer un counter a las frases
    const {counter, increment} = useEnumerator(1);

    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    // console.log(state);
    // console.log(loading);
    // console.log(data);

    // Si existe la data(true) extraer la posici√≥n 0 de data;
    const {author, quote} = !!data && data[0];
    // console.log(author, quote);

    return (
        <div>
            <h1>BreakingBad Quotes</h1>
            <hr/>

            {
                loading ? (
                    <div className='alert alert-info text-center'>
                        Loading...
                    </div>
                ) : (
                    <blockquote className='blockquote text-right'>
                        <p className='mb-0'>{quote}</p>
                        <footer className='blockquote-footer mt-1'>{author}</footer>
                    </blockquote>
                )
            }

            <button
                className='btn btn-primary'
                onClick={increment}>
                Siguiente frase
            </button>
        </div>
    )
}

export default MultipleCustomHook;