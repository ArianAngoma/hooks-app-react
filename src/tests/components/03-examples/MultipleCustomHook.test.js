import {shallow} from "enzyme";
import MultipleCustomHook from "../../../components/03-examples/MultipleCustomHook";
import useFetch from "../../../hooks/useFetch";
import useEnumerator from "../../../hooks/useEnumerator";

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useEnumerator');


describe('Test to component <MultipleCustomHoks/>', () => {
    beforeEach(() => {
        useEnumerator.mockReturnValue({
            counter: 10,
            increment: () => {
            }
        });
    })

    test('Should display the component <MultipleCustomHoks/> correctly', () => {
        useFetch.mockReturnValue({
            data: null, loading: true, error: null
        });

        const wrapper = shallow(<MultipleCustomHook/>);

        expect(wrapper).toMatchSnapshot();
    });

    test('Should display the information', () => {
        useFetch.mockReturnValue({
            data: [{author: 'Arian', quote: 'Hola Mundo'}], loading: false, error: null
        });

        const wrapper = shallow(<MultipleCustomHook/>);

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
        expect(wrapper.find('footer').text().trim()).toBe('Arian');
    })
})