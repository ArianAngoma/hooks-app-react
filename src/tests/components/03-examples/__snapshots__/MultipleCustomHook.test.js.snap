// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to component <MultipleCustomHoks/> Should display the component <MultipleCustomHoks/> correctly 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
